ADAL/src/request/ADAuthenticationRequest+Broker.m

```diff
@@ -186,74 +185,86 @@ + (ADAuthenticationResult *)processBrokerResponse:(NSURL *)response
     }
     
     // NSURLComponents resolves some URLs which can't get resolved by NSURL
-    NSURLComponents* components = [NSURLComponents componentsWithURL:response resolvingAgainstBaseURL:NO];
+    NSURLComponents *components = [NSURLComponents componentsWithURL:response resolvingAgainstBaseURL:NO];
     NSString *qp = [components percentEncodedQuery];
     //expect to either response or error and description, AND correlation_id AND hash.
-    NSDictionary* queryParamsMap = [NSDictionary adURLFormDecode:qp];
-    
+    NSDictionary *queryParamsMap = [NSDictionary adURLFormDecode:qp];
+
+    NSString *keychainGroup = resumeDictionary[@"keychain_group"];
+
     if([queryParamsMap valueForKey:OAUTH2_ERROR_DESCRIPTION])
     {
-        return [ADAuthenticationResult resultFromBrokerResponse:queryParamsMap];
+        // In the case where Intune App Protection Policies are required, the broker may send back the Intune MAM Resource token
+        NSMutableDictionary *brokerResponse = [[NSMutableDictionary alloc] initWithDictionary:queryParamsMap];
+        if (queryParamsMap[BROKER_INTUNE_HASH_KEY] && queryParamsMap[BROKER_INTUNE_RESPONSE_KEY])
+        {
+            ADAuthenticationError *intuneTokenError = nil;
+            NSDictionary *responseDictionary = @{BROKER_RESPONSE_KEY:queryParamsMap[BROKER_INTUNE_RESPONSE_KEY],
+                                                 BROKER_HASH_KEY:queryParamsMap[BROKER_INTUNE_HASH_KEY],
+                                                 BROKER_MESSAGE_VERSION:queryParamsMap[BROKER_MESSAGE_VERSION] ? queryParamsMap[BROKER_MESSAGE_VERSION] : @1};
+
+            NSDictionary *intuneTokenResponse = [ADBrokerKeyHelper decryptBrokerResponse:responseDictionary
+                                                                   correlationId:correlationId
+                                                                           error:&intuneTokenError];
+
+            ADAuthenticationResult *intuneTokenResult = [[ADTokenCacheItem new] processTokenResponse:intuneTokenResponse
+                                                                                    fromRefreshToken:nil
+                                                                                requestCorrelationId:intuneTokenResponse[OAUTH2_CORRELATION_ID_RESPONSE]];
+
+            if (!keychainGroup)
+            {
+                AD_LOG_WARN(correlationId, @"Failed to cache Intune token, unable to acquire keychain group.");
+            }
+            else if (AD_SUCCEEDED != intuneTokenResult.status)
+            {
+                AD_LOG_WARN(correlationId, @"Failed to acquire Intune token.");
+            }
+            else
+            {
+                ADTokenCacheAccessor *cacheAccessor = [[ADTokenCacheAccessor alloc] initWithDataSource:[ADKeychainTokenCache keychainCacheForGroup:keychainGroup]
```

Comment:
nit: what's the difference between this caching call and the one below? Can we combine those two?

Reply:
224-229 caches the Intune token if one was returned, but then discards intuneTokenResult and processes the broker response as an error, where 258-265 updates the success result and eventually returns it.  The cache access is the same, but they process the user differently.